File: leetcode/medium/3Sum.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        """
        Finds all unique triplets in the array which give the sum of zero.

        The solution uses a two-pointer approach after sorting the array, which is an
        efficient way to solve this problem with a time complexity of O(n^2).

        Algorithm:
        1.  Sort the input array `nums`. Sorting allows us to easily handle duplicates
            and use the two-pointer technique.
        2.  Iterate through the array with an index `i` from the start up to `len(nums) - 2`.
            The element `nums[i]` will be the first element of a potential triplet.
        3.  To avoid duplicate triplets, we skip any element `nums[i]` that is the same
            as the one before it.
        4.  For each `nums[i]`, initialize two pointers: `left = i + 1` and
            `right = len(nums) - 1`. These pointers will search for the other two
            elements of the triplet in the rest of the sorted array.
        5.  In a loop, calculate the sum `nums[i] + nums[left] + nums[right]`:
            - If the sum is less than 0, we need a larger sum, so we move the `left`
              pointer to the right (`left += 1`).
            - If the sum is greater than 0, we need a smaller sum, so we move the `right`
              pointer to the left (`right -= 1`).
            - If the sum is exactly 0, we have found a valid triplet. We add it to our
              result list.
        6.  After finding a valid triplet, we must move the pointers and skip any
            duplicate elements to ensure uniqueness. We increment `left` and decrement
            `right` while they point to the same values as before.

        Time Complexity: O(n^2). The O(n log n) for sorting is dominated by the O(n^2)
        of the nested loops (the outer `for` loop and the inner `while` loop).
        Space Complexity: O(1) or O(n), depending on the space complexity of the sorting
        algorithm used. The space for the result list is not counted.
        """
        # Sort the array to enable the two-pointer approach and handle duplicates easily.
        nums.sort()
        result = []
        # Iterate through the array. `i` will be the first element of the triplet.
        # We only need to go up to len(nums) - 2 to leave space for the two pointers.
        for i in range(len(nums) - 2):
            # *** Optimization: Skip duplicate first elements ***
            # If the current element is the same as the previous one, we've already
            # processed all possible triplets starting with that value, so we skip it.
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            # Initialize two pointers for the rest of the array.
            left, right = i + 1, len(nums) - 1
            while left < right:
                total = nums[i] + nums[left] + nums[right]
                # If the sum is too small, move the left pointer to a larger value.
                if total < 0:
                    left += 1
                # If the sum is too large, move the right pointer to a smaller value.
                elif total > 0:
                    right -= 1
                # If the sum is zero, we found a triplet.
                else:
                    result.append([nums[i], nums[left], nums[right]])
                    # *** Optimization: Skip duplicate second and third elements ***
                    # Move the left pointer forward as long as it's pointing to a duplicate.
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    # Move the right pointer backward as long as it's pointing to a duplicate.
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    # Move pointers to the next unique elements.
                    left += 1
                    right -= 1
        return result
        

File: leetcode/medium/3Sum.java

import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;

/**
 * Solution for the 3Sum problem.
 * The method finds all unique triplets in the array which give the sum of zero.
 * The overall time complexity is O(n^2) due to the nested loop structure, where the outer loop runs n times and the inner two-pointer loop runs n times in the worst case.
 * The space complexity is O(1) for the variables, and the space for the result list is not counted.
 */
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        // Sort the array to enable the two-pointer approach and easily handle duplicates.
        Arrays.sort(nums);
        int length = nums.length;
        List<List<Integer>> res = new ArrayList<>();

        // Iterate through the array, considering each element as a potential first element of a triplet.
        for (int i = 0; i < length - 2; i++) {
            // Skip duplicate elements for the first element of the triplet to avoid duplicate triplets.
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            // Use two pointers, one starting from the element after the current one, and one from the end of the array.
            int left = i + 1, right = length - 1;
            while (left < right) {
                int total = nums[i] + nums[left] + nums[right];

                // If the sum is less than zero, we need a larger sum, so move the left pointer to the right.
                if (total < 0) {
                    left++;
                }
                // If the sum is greater than zero, we need a smaller sum, so move the right pointer to the left.
                else if (total > 0) {
                    right--;
                }
                // If the sum is zero, we found a triplet.
                else {
                    res.add(List.of(nums[i], nums[left], nums[right]));
                    
                    // Skip duplicate elements for the second element of the triplet.
                    while (left < right && nums[left] == nums[left + 1]) {
                        left++;
                    }
                    
                    // Skip duplicate elements for the third element of the triplet.
                    while (left < right && nums[right] == nums[right - 1]) {
                        right--;
                    }

                    // Move both pointers to find the next potential triplet.
                    left++;
                    right--;
                }
            }
        }       
        return res;
    }
}